internal/hcp/internal/controllers/link/controller.go

```diff
@@ -80,3 +93,61 @@ func (r *linkReconciler) Reconcile(ctx context.Context, rt controller.Runtime, r
 
 	return nil
 }
+
+type linkInitializer struct {
+	cloudConfig config.CloudConfig
+}
+
+func (i *linkInitializer) Initialize(ctx context.Context, rt controller.Runtime) error {
+	if !i.cloudConfig.IsConfigured() {
+		return nil
+	}
+
+	// Construct a link resource to reflect the configuration
+	data, err := anypb.New(&pbhcp.Link{
+		ResourceId:   i.cloudConfig.ResourceID,
+		ClientId:     i.cloudConfig.ClientID,
+		ClientSecret: i.cloudConfig.ClientSecret,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Create the link resource for a configuration-based link
+	w := retry.Waiter{
+		MinWait: 1 * time.Second,
+		MaxWait: 15 * time.Second,
+		Jitter:  retry.NewJitter(50),
+	}
+	retryCtx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+	for {
+		_, writeErr := rt.Client.Write(ctx,
+			&pbresource.WriteRequest{
+				Resource: &pbresource.Resource{
+					Id: &pbresource.ID{
+						Name: types.LinkName,
+						Type: pbhcp.LinkType,
+					},
+					Metadata: map[string]string{
+						MetadataSourceKey: MetadataSourceConfig,
+					},
+					Data: data,
+				},
+			},
+		)
+		if writeErr != nil {
+			// Retry on error
+			rt.Logger.Trace("error initializing link, retrying", "error", writeErr)
+			if err := w.Wait(retryCtx); err != nil {
+				// Retry timed out, return last write error
+				return writeErr
+			}
+			continue
+		}
+		// Write was successful, break out of retry loop
+		break
+	}
```

Comment:
This was mostly to control what errors are logged! The resource write initially errors, but those errors that are essentially ignorable, like leader election hasn't occurred yet. With the internal retry, the user won't see any errors logged vs. with just the controller supervisor, there were a lot of errors logged that just eventually resolve themselves without needing any action from the user.

And then if the errors continue beyond timeout of the internal retry, then we do care about these errors since they're likely not the ignorable ones, so at that point, I want the controller supervisor to log an error.

Reply:
That is understandable and a reasonable trade-off. However, I worry that the two layers of retries ventures into making things more complex and harder to reason about in the longer term. 

Maybe there are other things we can reconsider to make this pattern an exception to the rule rather than a norm that other controller initializers might end up replicating.

Given most things V2 related are a significant departure architecturally (eventually do "the thing", just retry forever if it fails) vs V1 (do "the thing" now, block, or fail fast), it is probably worth exploring alternatives to smooth over the experience you're having.

Alternatives:
- consider making the retry strategy the controller supervisor uses configurable per controller
- tweak the logging that the supervisor retryer emits to be less distracting/more useful
- AFAIK, we have no concept of "open for business" - it means many different things based the context in which it is evaluated. Maybe codifying these as preconditions for controllers would prevent some of the "don't even bother trying to to start unless X, Y, and Z are true". Example preconditions:
  - a leader has been elected
  - the ACL subsystem has been bootstrapped if ACLs are enabled
  - builtin tenancy resources have been created and initialized
  - all or some of the above.

In any case, it would be ideal to not have nested layers of retries, but it might take a bit more operational experience with V2 controllers to be able to make useful progress in this area.

Thanks for taking this on and contributing to core functionality.

Reply:
That makes sense to me! I’ve removed the retry and went with having the Initialize method log errors rather than the controller’s run method (which was what was logging and not the controller supervisor like I had previously thought).

However, I really like the preconditions alternative since it addresses the real issue, which is that I need a certain condition met before I start the initialization. It would be nice too if the preconditions could be reusable across different controllers, as I imagine there will be common use cases like the ones you listed as examples.

I think it’ll take me some time to figure out how to properly check if a leader has been elected, though, just because of my limited knowledge of that part of the Consul code. Since this PR is blocking some dependent tasks, I went with adjusting the logging as a faster fix. I’ve created [a Jira](https://hashicorp.atlassian.net/browse/CC-7329) to follow up on this, though!